### DISCUSSION TITLE
Make a billiards game

### DISCUSSION THREAD

**Discussion by conant** [bepu/bepuphysics1#10] [Q&A]
Hi Nordby, 

Hope you are doing well.
I have made a topic about "Make a billiards game" here: https://forum.bepuentertainment.com/viewtopic.php?f=4&t=2709

With all your suggestions, this game is running well, but two problems:
1. It delays the simulation a bit at the first time I make a ApplyImpulse.
2. Sometimes, on some phone, with a low rate, It breaks determinism at some shot.

I believe there must be some static variables that cause the problems but I don't know how to find them.
They must be fixed before releasing the game, so I really need your help once more time.

**Comment by RossNordby** [comment]
>It delays the simulation a bit at the first time I make a ApplyImpulse.

Do you mean there's a spike in execution time? If so, it's probably the just in time compiler consuming a previously unseen codepath. You can explicitly visit troublesome codepaths at load time to avoid it, or force the JIT to compile them explicitly, or use some variant of pre-JITted (like NGEN) or AOT compilation.

>Sometimes, on some phone, with a low rate, It breaks determinism at some shot.

Sorry, there are too many possibilities to give a useful answer there. I can only offer the standard recommendations- try reproducing the problem in the simplest possible context, like a demo, so you can rule out most potential interference. Determinism is easily broken by even a single out of order interaction with the engine, so there's a lot of things it could be.

**Reply by conant** [reply]
> Do you mean there's a spike in execution time?

Yes I do. I don't think it's a JIT problem since I'm using Unity IL2CPP build system. However, this is currently solved by doing a trick that I call Space.Update() in dozen of frames ahead of making any ApplyImpulse.

> Sorry, there are too many possibilities to give a useful answer there. I can only offer the standard recommendations- try reproducing the problem in the simplest possible context, like a demo, so you can rule out most potential interference. Determinism is easily broken by even a single out of order interaction with the engine, so there's a lot of things it could be.

Thanks for your advice. I added logs to the code and tried to reproduce it successfully. The cause I found is not about the physics, It's actually a network sync issue. It's fixed now.

**Comment by conant** [comment]
Hi Nordby,

Recently, I have a bug that the balls fly out of the table (a shpere penetrating through a box) if ApplyImpulse an enough strong force to a ball (sphere) close to a rail (box).

I wonder if there is any configuration parameter to adjust to prevent that from happening?

**Reply by RossNordby** [reply]
Depends on the nature of the problem. If it's because of discrete timestepping, using a higher update rate or enabling CCD for the relevant objects would help. If it's constraint softness, you could reduce CollisionResponseSettings.Softness.

**Reply by conant** [reply]
Yep, using a higher update rate does help. I found that increasing box width fixes the problem too. Thanks!

**Comment by conant** [comment]
Hi Nordby,

I want to let you know that the game is in beta release phase with all your kindly supports. Thank you so much!

Currently, we only have a problem about user experience: the balls seem running short if compared with other billiards games.

I already reduced friction to minimal (e.g. 0.3) but it doesn't help. Is there any settings for me to tweak that?

Thanks!

**Reply by RossNordby** [reply]
Nice work!

> the balls seem running short if compared with other billiards games. 

My first guess would be damping, which directly removes energy from the body. The older thread mentioned damping values of 0.12 for linear and 0.45 for angular- those would definitely be enough to stop the balls fast. Reducing those to closer to zero would let things keep moving longer.

**Comment by conant** [comment]
Oh, I have overlooked these values. Thanks your guess!

**Comment by conant** [comment]
Currently, I have a problem with the speed of the game. The balls already move longer but slower. I manage to fix by increasing Space.Update(speed) but it's still a bit slow at the end of shot. I think it would be perfect if there's a config to make the balls move faster. For easier understanding, I give my current settings here:

`// Settings for billiards game`
`InertiaHelper.InertiaTensorScale = 1;`
`MotionSettings.DefaultPositionUpdateMode = PositionUpdateMode.Continuous;`
`CollisionResponseSettings.BouncinessVelocityThreshold = .1m;`
`SolverSettings.DefaultMinimumIterationCount = 0; // high speed`

`// Space settings`
`space.ForceUpdater.Gravity = new BVector3(0, 0, 10);`
`space.DeactivationManager.VelocityLowerLimit = 1m; // make game faster`
`space.Solver.IterationLimit = 5; // high speed`
`space.TimeStepSettings = new BEPUphysics.TimeStepSettings`
`{`
`    TimeStepDuration = 1 / 100m, // for more accurate physics, fix some abnormal bugs`
`    MaximumTimeStepsPerFrame = 10`
`};`

`// Update Space each frame`
`space.Update(.05m); // make game faster`

Any help would be appreciated!

**Reply by RossNordby** [reply]
At the moment, you're using the internal timestepping overload of `Space.Update`. Given the time step duration, that is equivalent to calling `space.Update()` without a parameter 5 times. If the amount of time you want to simulate is exactly 5/100 every call, then this is fine- for example, if you have an external timing loop that ensures that it is called once every 5/100 units of time in real time, then that can work.

If `space.Update` is being called from a context where the time passed since the last call isn't exactly the same, then it'll cause odd behavior. For example, if the rate of the timestepping is linked to the rate of rendered frames, then higher FPS will yield faster simulation, which is usually not desired. In this situation, you'll want to timestep a variable number of times based on the elapsed time.

**Reply by conant** [reply]
It really helps! I haven't fully understood Space.Update until reading your comment. After applying your suggestion, means calling Space.Update() in FixedUpdate (constant elapsed time), the game seems running the same in every platforms. There's only one issue left that it seems not balancing between physics speed and game speed. Like the above my question, for the balls to finish moving faster, I must increase the physics speed to a high value that makes the balls move too fast at the beginning of shot (after ApplyImpulse with a strong force). Is there any tool for me to customize in this situation?

**Reply by RossNordby** [reply]
I'm not sure exactly what behavior you're observing, or looking for, so some general advice:

To start with, make sure your values for all velocities and timestep durations are physically consistent (for example, use measured values from a real billiards game), and ensure that the timesteps are being executed as quickly as they should be in real time.

Once you have that foundation, you can tune stuff like damping and friction to get closer to your desired behavior.

**Reply by conant** [reply]
After tweaking many times, I found some settings that may be reasonable for this game. Once more time, thanks for all your help and support and for creating the amazing library. I wish you all the best and would stay strong through the pandemic times :)

**Reply by RadiantSolarSolutions** [reply]
Edit 2: Well, Spheres seem to work okay, but Convex Hulls and anything slightly more complex seems to have issues. They either have endless jittering/jiggling or get stuck inside one another or both. Tried a lot of different settings with no luck.
